### PR TITLE
Progress service for redeem and unstake

### LINKED ISSUE #153: Progress redeem service

### PR DESCRIPTION
**Pull request by abhayks1**
PR covers below:

- Implementation of progress service which performs redeem and unstake operations.
- MessageRepository is attached with ProgressService

Fixes #153

### CODE REVIEW

File: test/services/redeem_and_unstake/ProgressService/update.test.ts [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+  let progressRedeemSpy: any;
+  let progressUnstakeSpy: any;
+  let utilsSpy: any;
+  let gatewayRepository: any;
+  let eip20CoGatewayMockObject: any;
+  let progressService: ProgressService;
+  let message: any;
+  const originWeb3 = new Web3(null);
+  const auxiliaryWeb3 = new Web3(null);
+  const originWorkerAddress = '0xA1e801AbF4288a38FfFEa3084C826B810c5d5294';
```
Comment (puneet-khushwani-eth) [inline test/services/redeem_and_unstake/ProgressService/update.test.ts]: any specific reasons for not having dummy values for origin and aux worker addresses ? As sendTx is mocked dummy values might work here

File: test/services/redeem_and_unstake/ProgressService/update.test.ts [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+      'progressUnstake',
+      sinon.fake.returns(progressUnstakeRawTx),
+    );
+    gatewaySpy = sinon.replace(
+      interacts,
+      'getEIP20Gateway',
+      sinon.fake.returns(eip20GatewayMockObject as any),
+    );
+
+    utilsSpy = sinon.replace(
```
Comment (puneet-khushwani-eth) [inline test/services/redeem_and_unstake/ProgressService/update.test.ts]:
```suggestion
    utilsSendTxSpy = sinon.replace(
```

File: test/services/redeem_and_unstake/ProgressService/update.test.ts [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+      1,
+      [[originWeb3, gatewayAddress]],
+    );
+
+    sinon.restore();
+  });
+
+  it('should not progress on origin and auxiliary when source status '
+    + 'is not declared', async () => {
+    message.sourceStatus = MessageStatus.Undeclared;
```
Comment (puneet-khushwani-eth) [inline test/services/redeem_and_unstake/ProgressService/update.test.ts]: when we add direction checks in service, we should also add some more negative permutations in test cases

File: src/services/Services.ts [hunk 4]
```diff
@@ -19,27 +20,32 @@ export default class Services {
...
 
   public readonly proveCoGatewayService: ProveCoGatewayService;
 
   /**
    * @param acceptStakeRequestService Instance of accept stake request service.
    * @param proveGatewayService Instance of prove gateway service.
    * @param confirmStakeIntentService Instance of confirm stake intent service.
-   * @param progressService Instance of progress service.
+   * @param stakeProgressService Instance of progress service.
+   * @param proveCoGatewayService Instance of prove cogateway service.
```
[the remark this answers is not in the document]
Comment (abhayks1) [reply src/services/Services.ts]: I guess you meant add `redeemProgressService` to documentation 😄 
Done!

File: src/services/Services.ts [hunk 5]
```diff
@@ -19,27 +20,32 @@ export default class Services {
...
+  public readonly redeemProgressService: RedeemProgressService;
 
   public readonly proveCoGatewayService: ProveCoGatewayService;
 
   /**
    * @param acceptStakeRequestService Instance of accept stake request service.
    * @param proveGatewayService Instance of prove gateway service.
    * @param confirmStakeIntentService Instance of confirm stake intent service.
-   * @param progressService Instance of progress service.
+   * @param stakeProgressService Instance of progress service.
```
[the remark this answers is not in the document]
Comment (abhayks1) [reply src/services/Services.ts]: Done!

File: src/services/redeem_and_unstake/ProgressService.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    const eip20CoGateway = interacts.getEIP20CoGateway(
+      this.auxiliaryWeb3,
+      this.coGatewayAddress,
+    );
+    const transactionOptions = {
+      from: this.auxWorkerAddress,
+      gasPrice: AUXILIARY_GAS_PRICE,
+    };
+
+    assert(message.secret !== undefined, 'message secret is undefined');
```
[the remark this answers is not in the document]
Comment (abhayks1) [reply src/services/redeem_and_unstake/ProgressService.ts]: Changes done!

File: src/services/redeem_and_unstake/ProgressService.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+  /**
+   * This method react on changes in Message model.
+   * @param messages List of Message models.
+   */
+  public async update(messages: Message[]): Promise<void> {
+    Logger.debug('Progress service invoked');
+
+    const progressPromises = messages
+      .filter(message => message.isValidSecret())
```
[the remark this answers is not in the document]
Comment (abhayks1) [reply src/services/redeem_and_unstake/ProgressService.ts]: Added filter for message.type which differentiates between stake and redeem.

File: test/services/redeem_and_unstake/ProgressService/update.test.ts (hunk 1 above)
Comment (abhayks1) [reply test/services/redeem_and_unstake/ProgressService/update.test.ts]: I don't think we have strict requirement. Anyways I have updated the addresses.

File: test/services/redeem_and_unstake/ProgressService/update.test.ts (hunk 2 above)
Comment (abhayks1) [reply test/services/redeem_and_unstake/ProgressService/update.test.ts]: Done!

File: test/services/redeem_and_unstake/ProgressService/update.test.ts (hunk 3 above)
Comment (abhayks1) [reply test/services/redeem_and_unstake/ProgressService/update.test.ts]: Done!

**Comment by abhayks1** [comment]
Available for review again 🐰

File: src/services/redeem_and_unstake/ProgressService.ts [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+import Web3 from 'web3';
+
+import { interacts } from '@openst/mosaic-contracts';
+
+import { AUXILIARY_GAS_PRICE, ORIGIN_GAS_PRICE } from '../../Constants';
+import Logger from '../../Logger';
+import Gateway from '../../models/Gateway';
+import Message from '../../models/Message';
+import GatewayRepository from '../../repositories/GatewayRepository';
+import {MessageStatus, MessageType} from '../../repositories/MessageRepository';
```
Comment (puneet-khushwani-eth) [inline src/services/redeem_and_unstake/ProgressService.ts]:
```suggestion
import { MessageStatus, MessageType } from '../../repositories/MessageRepository';
```

File: src/services/stake_and_mint/ProgressService.ts [hunk 9]
```diff
@@ -8,7 +8,7 @@ import Logger from '../../Logger';
 import Gateway from '../../models/Gateway';
 import Message from '../../models/Message';
 import GatewayRepository from '../../repositories/GatewayRepository';
-import { MessageStatus } from '../../repositories/MessageRepository';
+import {MessageStatus, MessageType} from '../../repositories/MessageRepository';
```
Comment (puneet-khushwani-eth) [inline src/services/stake_and_mint/ProgressService.ts]:
```suggestion
import { MessageStatus, MessageType } from '../../repositories/MessageRepository';
```

File: src/services/stake_and_mint/ProgressService.ts [hunk 10]
```diff
@@ -59,6 +59,7 @@ export default class ProgressService {
     Logger.debug('Progress service invoked');
 
     const progressPromises = messages
+      .filter(message => message.type === MessageType.Stake)
```
Comment (puneet-khushwani-eth) [inline src/services/stake_and_mint/ProgressService.ts]: both filters can be merged to save the extra loop.

File: test/services/redeem_and_unstake/ProgressService/update.test.ts [hunk 11]
```diff
@@ -0,0 +1,387 @@
+import sinon from 'sinon';
+import Web3 from 'web3';
+
+import { interacts } from '@openst/mosaic-contracts';
+
+import { AUXILIARY_GAS_PRICE, ORIGIN_GAS_PRICE } from '../../../../src/Constants';
+import Message from '../../../../src/models/Message';
+import GatewayRepository from '../../../../src/repositories/GatewayRepository';
+import {MessageStatus, MessageType} from '../../../../src/repositories/MessageRepository';
```
Comment (puneet-khushwani-eth) [inline test/services/redeem_and_unstake/ProgressService/update.test.ts]:
```suggestion
import { MessageStatus, MessageType } from '../../../../src/repositories/MessageRepository';
```

File: src/services/redeem_and_unstake/ProgressService.ts [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+  /**
+   * This method react on changes in Message model.
+   * @param messages List of Message models.
+   */
+  public async update(messages: Message[]): Promise<void> {
+    Logger.debug('Progress service invoked');
+
+    const progressPromises = messages
+      .filter(message => message.type === MessageType.Redeem)
+      .filter(message => message.isValidSecret())
```
Comment (puneet-khushwani-eth) [inline src/services/redeem_and_unstake/ProgressService.ts]: both filters can be merged to save the extra loop.

File: src/services/stake_and_mint/ProgressService.ts (hunk 9 above)
Comment (abhayks1) [reply src/services/stake_and_mint/ProgressService.ts]: Done!

File: src/services/stake_and_mint/ProgressService.ts (hunk 10 above)
Comment (abhayks1) [reply src/services/stake_and_mint/ProgressService.ts]: Done!

File: test/services/redeem_and_unstake/ProgressService/update.test.ts (hunk 11 above)
Comment (abhayks1) [reply test/services/redeem_and_unstake/ProgressService/update.test.ts]: Done!

File: src/services/redeem_and_unstake/ProgressService.ts (hunk 12 above)
Comment (abhayks1) [reply src/services/redeem_and_unstake/ProgressService.ts]: Done!

**Pull request merged**